<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>16S Community Analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Edward Lopatto</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Front Door</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="community_tutorial.html">16s Community Analysis</a>
    </li>
    <li>
      <a href="str_tutorial.html">str Gene Variant Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="helpful_references.html">Helpful References</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">16S Community Analysis</h1>
<h4 class="date"><em>Last updated 7/23/2018</em></h4>

</div>


<div id="background" class="section level1">
<h1>Background</h1>
<p>I used this workflow to analyze 16s Miseq data from the Grinnell CAFO project, but this tutorial should be applicable to any 16s data.</p>
<p>In order to better understand this tutorial, you will need a basic understanding of bash, and R. If you are want to learn either or need a review, see the helpful references tab. I try to show all of the details, but I may accidentally gloss over rudimentary skills. You should be able to do a quick google search and find what you need in that case. If not, feel free to ask me.</p>
<p>You will also need the following installed on an AWS instance(see MOTHUR hints for instance storage details) or HPC (See references for help):</p>
<ul>
<li>MOTHUR (There is a <a href="https://mothur.org/wiki/Mothur_AMI">preloaded AWS instance</a> that is convenient, but you can also load a fresh MOTHUR)</li>
</ul>
<p>You will also need the following installed on your own computer (See references for help):</p>
<ul>
<li><a href="https://filezilla-project.org/">FileZilla</a> or know another way to transfer files between computers</li>
<li><a href="https://www.rstudio.com/">Rstudio</a></li>
</ul>
<p>So you have 16s sequences from the Miseq and the programs above set up, lets begin. If you don’t have 16s sequences yet, the preloaded mothur AMI has sequences you can use.</p>
</div>
<div id="mothur" class="section level1">
<h1>MOTHUR</h1>
<div id="prepping-mothur" class="section level2">
<h2>Prepping MOTHUR</h2>
<p>If you are using the preloaded AMI, delete all the files in the ~/data/raw directory. If you aren’t <a href="https://github.com/mothur/mothur_ami/blob/master/README.md">these instructions</a> may help setting up mothur.</p>
<p>Head into the ~/data/references directory and make sure your references are up to date. We will be using the <a href="https://mothur.org/wiki/Silva_reference_files">SILVA</a> (use the full NR database) and <a href="https://mothur.org/wiki/RDP_reference_files">rdp</a> references files. You can delete the HMP_mock file, and upload your own mock community reference to the ~/data/references directory.</p>
<pre><code>#You can upload the references directly to the server using the command wget. This saves your own computer lots of space. Make sure to update the reference links in the commands.

#https://mothur.org/wiki/Silva_reference_files
wget -P data/references https://mothur.org/w/images/a/a4/Silva.seed_v128.tgz
tar xvzf data/references/Silva.seed_v128.tgz -C data/references
rm data/references/Silva.seed_v128.tgz data/references/README.*

#https://mothur.org/wiki/RDP_reference_files
wget -P data/references https://mothur.org/w/images/c/c3/Trainset16_022016.pds.tgz
tar xvzf data/references/Trainset16_022016.pds.tgz -C data/references
rm data/references/Trainset16_022016.pds.tgz
mv data/references/trainset16_022016.pds/* data/references
rm -rf data/references/trainset16_022016.pds data/references/README.*</code></pre>
<p>Using FileZilla, transfer the sequences (including the mock community if you have it) into the ~/data/raw directory.</p>
<p>After the references are updated and sequences uploaded, MOTHUR should be prepped and ready to go.</p>
</div>
<div id="running-mothur" class="section level2">
<h2>Running MOTHUR</h2>
<p>I thought about writing an original tutorial for using MOTHUR, but it would be redundant. Pat Schloss and co. made an <a href="https://mothur.org/wiki/EDAMAME">amazing tutorial</a> and now that you are set up, you should be able to follow the tutorial exactly. Using the defaults should work well for our purposes. For this tutorial, you will only need to follow the tutorial through to the “Analysis” section, where you rename files right before subsampling, but you are welcome to complete the MOTHUR tutorial. Once you have files called “stability.opti_mcc.shared” and “stability.cons.taxonomy” you are good to move on to the next section.</p>
<p>However, I will give you some helpful hints, and solutions to problems I encountered.</p>
<ul>
<li>File names can be what ever you want them to be. They don’t need to end in “.stability”. Make sure they make sense to you and are easy to keep track of. Ideally, they also would make sense to other people, too.</li>
<li>Warnings are just warnings. Be aware of them, but it’s probably ok to continue on. Errors will not let you continue.</li>
<li>Don’t just blindly follow the tutorial. Understand what each step does and pay attention. Sometimes, the defaults won’t work or throw away too many sequences!</li>
<li>I recommend doing the analysis in one session (might require skill with <a href="https://github.com/edamame-course/2015-tutorials/blob/master/final/2015-06-22_tmux.md">tmux</a>), but you can do MOTHUR in multiple sessions. If you do multiple sessions, make sure you are calling the right directories (I used <a href="https://mothur.org/wiki/Set.dir">set.dir</a> a lot).</li>
</ul>
<p>If at some point MOTHUR is not behaving and you get stuck, be mindful of the following:</p>
<ul>
<li>Try going back and redoing the commands from the previous section. MOTHUR requires steps be in exact order, and its possible you missed a prior command or did not work properly.</li>
<li>MOTHUR takes up a huge amount of space. It outputs lots of text files per a sequence, and each sequence is already taking up a large amount of memory. Try restarting on a larger instance with more memory. To analyze about 150 sequences, I had to use a m4.4xlarge and added on even more storage. <strong>NOTE: THE LARGER THE INSTANCE THE MORE IT COSTS. (update 7/24/18) AMAZON ALSO JUST CHANGED THEIR POLICY. YOU NEED TO ASK PERMISSION TO USE LARGER INSTANCES. PLAN AHEAD, BECAUSE IT MAY TAKE TIME TO OBTAIN.</strong></li>
<li>The <a href="https://www.mothur.org/forum/">MOTHUR forums</a> can be helpful, but they are hard to search.</li>
<li>Use google as a last resort.</li>
<li>Email <a href="mailto:lopatto@grinnell.edu">lopatto@grinnell.edu</a> with the log file and description of the problem. I may be able to help you, but if I’m not, I have access to people who know more than me.</li>
</ul>
<p><strong>Make sure to cite: Kozich JJ, Westcott SL, Baxter NT, Highlander SK, Schloss PD. (2013): Development of a dual-index sequencing strategy and curation pipeline for analyzing amplicon sequence data on the MiSeq Illumina sequencing platform. Applied and Environmental Microbiology. 79(17):5112-20.</strong></p>
<p>Once you have the file called “stability.opti_mcc.shared” and “stability.cons.taxonomy”, save it to your personal computer and you are ready to move on.</p>
</div>
</div>
<div id="r-analysis-phyloseq" class="section level1">
<h1>R Analysis (Phyloseq)</h1>
<div id="loading-and-prepping-the-data" class="section level2">
<h2>Loading and prepping the data</h2>
<p>So now that you have “stability.opti_mcc.shared” and “stability.cons.taxonomy”, you should boot up Rstudio and are ready for analysis. For the purposes of this tutorial, I will show mostly everything that doesn’t take up too much space - both the r code input (in gray) and raw output (in white) including warning messages. A lot of this is modified from the <a href="https://joey711.github.io/phyloseq/tutorials-index.html">Phyloseq tutorial</a>.</p>
<p>First, load the libraries. Some of these may be redundant, but thats a problem to solve at a later date. There will be lots of warnings too, but it should be fine to ignore them.</p>
<pre class="r"><code>library(ape)
library(grid)
library(plyr)
library(reshape)
library(reshape2)
library(ggthemes)
library(gridExtra)
library(corrplot)
library(dplyr)
library(vegan)
library(SpadeR)
library(ggplot2)

library(RColorBrewer)
library(phyloseq)

library(DESeq2)</code></pre>
<p>Lets import the data. You should also make and import a meta data .csv file. The meta file will contain details about each sample (ex. temperature and date).</p>
<pre class="r"><code>#Importing output from MOTHUR
darte_ed_16s &lt;- import_mothur(mothur_shared_file = &quot;~/Desktop/stability.opti_mcc.shared&quot;, mothur_constaxonomy_file = &quot;~/Desktop/stability.cons.taxonomy&quot;)
darte_ed_16s</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19061 taxa and 156 samples ]
## tax_table()   Taxonomy Table:    [ 19061 taxa by 6 taxonomic ranks ]</code></pre>
<pre class="r"><code>#Importing metadata
metadata &lt;- read.csv(&quot;~/Desktop/16slabels.csv&quot;, row.names=1)
head(metadata)</code></pre>
<pre><code>##                   Day Type Site Sample. Replicate Depth ID             DT
## AH37POBT1S1R1 11/8/16 Soil    1       1         1   low  1 11/8/2016 Soil
## AH38POBT1S1R2 11/8/16 Soil    1       1         2   low  2 11/8/2016 Soil
## AH39POBT1S1R3 11/8/16 Soil    1       1         3   low  3 11/8/2016 Soil
## AH40POBT1S1R1 11/8/16 Soil    1       2         1   low  4 11/8/2016 Soil
## AH41POBT1S1R2 11/8/16 Soil    1       2         2   low  5 11/8/2016 Soil
## AH42POBT1S1R3 11/8/16 Soil    1       2         3   low  6 11/8/2016 Soil</code></pre>
<pre class="r"><code>sample &lt;- sample_data(metadata)
sample_data(darte_ed_16s)&lt;- sample

darte_ed_16s</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19061 taxa and 155 samples ]
## sample_data() Sample Data:       [ 155 samples by 8 sample variables ]
## tax_table()   Taxonomy Table:    [ 19061 taxa by 6 taxonomic ranks ]</code></pre>
<p>Our data is messy. It’s not rarefied. It has NTC sequences in it. It has lots of small taxa with very low abundances. The following will clean your data in various ways, so you can graph and analyze it in many ways. We won’t use all of the cleaned data types in this tutorial, but it gives you options to think about. Make sure to change details, like your definition of too small taxa abundance, to fit your needs.</p>
<p><em>Common question: What depth do I rarefy at?</em> <em>Answer: Well there isn’t a right answer. Currently, I’ve heard rarefying at 10,000 sequences is ideal, and no less than 5,000 sequences. You can make a rarefaction curve in MOTHUR and plot it in R, to make a more precise decision. But, I like Pat Schloss’s advice: If you have crappy data (rarefying at anywhere near 5,000 sequences eliminates too much of your data to answer your questions), redo the sequencing, sequence more replicates, or get more DNA. If you can’t, rarefy at the lowest depth that answers your questions. Pat says the lowest he has gone is about 500 sequences. This is OK, but you need to be aware that this alters the picture of your data, and acknowledge that in the write up.</em></p>
<p>Consult others and try different rarefying depths to see if the picture alters. For this tutorial I use 400, because I’m working with bad data.</p>
<pre class="r"><code># set rarefying depth
after_remove_low_depth &lt;- prune_samples(sample_sums(darte_ed_16s) &gt;= 400, darte_ed_16s)
head(sample_sums(after_remove_low_depth))</code></pre>
<pre><code>## AH105OBT1S2R3 AH106OBT1S2R1      AH109MT2      AH110MT2      AH111MT2 
##           515           619         22510         24628         23400 
##      AH112MT2 
##         19082</code></pre>
<pre class="r"><code>set.seed(1)
rare &lt;- rarefy_even_depth(after_remove_low_depth, sample.size = 400,rngseed=TRUE)</code></pre>
<pre><code>## `set.seed(TRUE)` was used to initialize repeatable random subsampling.</code></pre>
<pre><code>## Please record this for your records so others can reproduce.</code></pre>
<pre><code>## Try `set.seed(TRUE); .Random.seed` for the full vector</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## 15128OTUs were removed because they are no longer 
## present in any sample after random subsampling</code></pre>
<pre><code>## ...</code></pre>
<pre class="r"><code>head(sample_sums(rare))</code></pre>
<pre><code>## AH105OBT1S2R3 AH106OBT1S2R1      AH109MT2      AH110MT2      AH111MT2 
##           400           400           400           400           400 
##      AH112MT2 
##           400</code></pre>
<pre class="r"><code>#remove the NTC sample. Check to make sure it doesn&#39;t have too many sequences before you through it away though! NTC and blanks are quality controls.
to_remove &lt;- c(&quot;NTC&quot;)
pruned &lt;- prune_samples(!(rownames(sample_data(rare)) %in% to_remove), rare)

#filter out OTUs less than 10
#darte_ed_16s_filter &lt;- filter_taxa(pruned, function(x) sum(x) &gt; 10, TRUE)

 #relative abundance
darte_ed_16s_filter_re &lt;- transform_sample_counts(pruned, function(x) x /sum(x))

#Get rid of small taxa
darte_ed_16s_filter2 &lt;- filter_taxa(darte_ed_16s_filter_re, function(x) sum(x) &gt; .001, TRUE)

#Combine OTUs with common taxa
darte_ed_16s_filter_re_g = tax_glom(darte_ed_16s_filter2, &quot;Rank2&quot;)</code></pre>
</div>
</div>
<div id="figures-and-statistics-in-r" class="section level1">
<h1>Figures and statistics in R</h1>
<p>Here are some figures and statistical examples that can be tweaked to your preferences.</p>
<div id="stacked-bar-figures" class="section level2">
<h2>Stacked Bar Figures</h2>
<p>This will give you a stacked bar of taxonomy composition (in this case phyla) for each sample. I also facetted by type. If you want by Genus set fill=“Rank3”. (Or just rename the ranks like a sensible person unlike me)</p>
<pre class="r"><code>#stacked bar by sample
plot_bar(darte_ed_16s_filter_re_g, fill=&quot;Rank2&quot;, facet_grid= &quot;.~Type&quot;) + scale_fill_hue()</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>You can also display select individual samples</p>
<pre class="r"><code>plot_bar(darte_ed_16s_filter_re_g, fill=&quot;Rank2&quot;) + scale_fill_hue() +scale_x_discrete(limits=c(&quot;AH110MT2&quot;))</code></pre>
<pre><code>## Warning: Removed 1656 rows containing missing values (position_stack).</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>This will give you the average taxa composition of each treatment type. I borrowed and edited the function from Jin Choi of the GERMS lab at Iowa State University.</p>
<pre class="r"><code>#Jin&#39;s stacked bar function
# You need to change &quot;DT&quot; into your name of treatment
group_bar_chart &lt;- function(physeq){
  fin_table  = data.frame()
  for (group in unique(sample_data(physeq)$DT) ){
    temp_physeq = prune_samples( (sample_data(physeq)$DT == group), physeq)

    new_table &lt;- data.frame(taxa_sums(temp_physeq), tax_table(temp_physeq)[,2])
    colnames(new_table) = c(&quot;abundance&quot;, &quot;phylum&quot;)
    sum_table = data.frame()
    for (x in unique(new_table$phylum) ){
      temp = subset(new_table, phylum==x)
      su = sum(temp$abundance)
      sum_table = rbind(sum_table, data.frame(su, temp[1,2]))
    }

    colnames(sum_table) = c(&quot;abundance&quot;, &quot;phylum&quot;)
    perc_table = sum_table
    for (i in 1:nrow(sum_table)){
      perc_table[i,1] = sum_table[i,1] / sum(sum_table[,1])
    }

    other_table = data.frame()
    other = c()
    for (i in 1:nrow(perc_table)){
      if(perc_table[i,1] &gt; 0.01) {
        other_table = rbind(other_table, perc_table[i,])
      }else{
        other = c(other, perc_table[i,1])
      }
    }

    sum(other)
    tep = data.frame(sum(other), &quot;other&quot;)
    colnames(tep) = c(&quot;abundance&quot;, &quot;phylum&quot;)
    tfin = rbind(other_table, tep)
    ttfin = cbind(tfin,group)
    fin_table = rbind(fin_table, ttfin)
    phy = unique(fin_table$phylum)
  }
  return(fin_table)
}


#Plot stacked bar. Change levels to your treatment identifications.
fin_table &lt;- group_bar_chart(pruned)
fin_table$group = factor(fin_table$group, levels = c(&quot;11/8/2016 Soil&quot;,&quot;11/12/2016 Manure&quot; ,&quot;11/15/2016 Manure Line&quot;,&quot;11/15/2016 Soil&quot;))
(p &lt;- ggplot(fin_table, aes(x=group,y=abundance, fill=phylum))+geom_bar(stat=&quot;identity&quot;,color=&quot;black&quot;)+labs(y=&quot;relative abundance&quot;)+ scale_fill_hue()+theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)))</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="boxplot-by-average-phyla" class="section level2">
<h2>Boxplot by average phyla</h2>
<p>Ok, you caught me. When you average taxa, there should be some display of variance. If we were to add error bars to Jin’s stacked bar plot, it would be too messy. The solution is to plot by taxa. This type of figure can be a good compliment to Jin’s stacked bar and appear in the supplemental info. Alternatively, you could use this figure as your main composition comparison figure. I only plotted the top 10 phyla.</p>
<pre class="r"><code>tax &lt;- tax_glom(darte_ed_16s_filter_re_g, taxrank=&quot;Rank2&quot;)
phylum10 = names(sort(taxa_sums(tax), TRUE)[1:10])
top10=prune_taxa(phylum10, tax)
dat &lt;- psmelt(top10)
ggplot(dat, aes(Type, Abundance))  + facet_wrap(~Rank2, ncol=5, scales= &quot;free&quot;) + geom_boxplot(aes(x = Type, y = Abundance, fill= Type))+theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="nmds-ordination" class="section level2">
<h2>NMDS Ordination</h2>
<p>Ordinations are good to show differences spatially between treatments. I use NMDS because it uses the least assumptions and is versatile, but you could alternatively use PCoA or another ordination method.</p>
<p>Pairwise.adonis is the statistical method I used to analyze and produce the NMDS data. The function is from Jin. Adonis is comparable to ANOSIM, just more versatile (This is a point you will probably need to clarify with reviewers, as ANOSIM is more common).</p>
<pre class="r"><code>#statistic method
pairwise.adonis &lt;- function(x,factors, sim.function = &#39;vegdist&#39;, sim.method = &#39;bray&#39;, p.adjust.m =&#39;bonferroni&#39;)
{

co = combn(unique(as.character(factors)),2)
pairs = c()
F.Model =c()
R2 = c()
p.value = c()

for(elem in 1:ncol(co)){
if(sim.function == &#39;daisy&#39;){
library(cluster); x1 = daisy(x[factors %in% c(co[1,elem],co[2,elem]),],metric=sim.method)
} else{x1 = vegdist(x[factors %in% c(co[1,elem],co[2,elem]),],method=sim.method)}

ad = adonis(x1 ~ factors[factors %in% c(co[1,elem],co[2,elem])] , permutations=9999);
pairs = c(pairs,paste(co[1,elem],&#39;vs&#39;,co[2,elem]));
F.Model =c(F.Model,ad$aov.tab[1,4]);
R2 = c(R2,ad$aov.tab[1,5]);
p.value = c(p.value,ad$aov.tab[1,6])
}
p.adjusted = p.adjust(p.value,method=p.adjust.m)
sig = c(rep(&#39;&#39;,length(p.adjusted)))
sig[p.adjusted &lt;= 0.05] &lt;-&#39;.&#39;
sig[p.adjusted &lt;= 0.01] &lt;-&#39;*&#39;
sig[p.adjusted &lt;= 0.001] &lt;-&#39;**&#39;
sig[p.adjusted &lt;= 0.0001] &lt;-&#39;***&#39;

pairw.res = data.frame(pairs,F.Model,R2,p.value,p.adjusted,sig)
print(&quot;Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1&quot;)
return(pairw.res)
}

#NMDS
data.selected &lt;- t(otu_table(pruned))
mds.all=metaMDS(data.selected,distance=&quot;bray&quot;, k=2)</code></pre>
<pre><code>## Square root transformation
## Wisconsin double standardization
## Run 0 stress 0.07903614 
## Run 1 stress 0.07904129 
## ... Procrustes: rmse 0.001633319  max resid 0.01159598 
## Run 2 stress 0.0790396 
## ... Procrustes: rmse 0.001430875  max resid 0.009460468 
## ... Similar to previous best
## Run 3 stress 0.08009188 
## Run 4 stress 0.07903485 
## ... New best solution
## ... Procrustes: rmse 0.00129612  max resid 0.007691899 
## ... Similar to previous best
## Run 5 stress 0.08081018 
## Run 6 stress 0.07903733 
## ... Procrustes: rmse 0.001120355  max resid 0.00783705 
## ... Similar to previous best
## Run 7 stress 0.07903751 
## ... Procrustes: rmse 0.001185811  max resid 0.007682485 
## ... Similar to previous best
## Run 8 stress 0.07903675 
## ... Procrustes: rmse 0.001038581  max resid 0.007739903 
## ... Similar to previous best
## Run 9 stress 0.08913522 
## Run 10 stress 0.07903939 
## ... Procrustes: rmse 0.002578773  max resid 0.01755279 
## Run 11 stress 0.07903803 
## ... Procrustes: rmse 0.001223159  max resid 0.007635474 
## ... Similar to previous best
## Run 12 stress 0.07903668 
## ... Procrustes: rmse 0.001044005  max resid 0.007732396 
## ... Similar to previous best
## Run 13 stress 0.0807317 
## Run 14 stress 0.08984432 
## Run 15 stress 0.08984387 
## Run 16 stress 0.07903483 
## ... New best solution
## ... Procrustes: rmse 9.213799e-05  max resid 0.000299594 
## ... Similar to previous best
## Run 17 stress 0.07903616 
## ... Procrustes: rmse 0.0006419224  max resid 0.003630855 
## ... Similar to previous best
## Run 18 stress 0.08009756 
## Run 19 stress 0.07950962 
## ... Procrustes: rmse 0.005514771  max resid 0.03343227 
## Run 20 stress 0.08946405 
## *** Solution reached</code></pre>
<pre class="r"><code>data.sm=as.data.frame(scores(mds.all))
data.sm$DT &lt;- as.factor(sample_data(rare) $DT)
pairwise.adonis(data.selected, as.factor(sample_data(rare) $DT))</code></pre>
<pre><code>## [1] &quot;Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1&quot;</code></pre>
<pre><code>##                                         pairs   F.Model         R2 p.value
## 1         11/8/2016 Soil vs 11/12/2016 Manure 19.785607 0.39741620  0.0001
## 2    11/8/2016 Soil vs 11/15/2016 Manure Line  1.022050 0.04641030  0.3974
## 3           11/8/2016 Soil vs 11/15/2016 Soil  3.288120 0.05454010  0.0001
## 4 11/12/2016 Manure vs 11/15/2016 Manure Line 21.870345 0.70845807  0.0933
## 5        11/12/2016 Manure vs 11/15/2016 Soil 25.702140 0.36352704  0.0001
## 6   11/15/2016 Manure Line vs 11/15/2016 Soil  0.885524 0.02400736  0.5777
##   p.adjusted sig
## 1     0.0006  **
## 2     1.0000    
## 3     0.0006  **
## 4     0.5598    
## 5     0.0006  **
## 6     1.0000</code></pre>
<pre class="r"><code>#plot
ggplot(data=data.sm, aes(x=NMDS1, y=NMDS2, color=DT)) + geom_point(size=5) +
geom_hline(yintercept=0.0, colour=&quot;grey&quot;, lty=2)+
geom_vline(xintercept=0.0, colour=&quot;grey&quot;,lty=2) +
scale_color_brewer(palette=&quot;Set1&quot;) + theme(legend.position = c(0.11, 0.9)) +
theme(legend.background = element_rect(fill=&quot;white&quot;, size=0.3, linetype=&quot;solid&quot;, colour=&quot;black&quot;))+stat_ellipse()+labs(color=&quot;Day&quot;)</code></pre>
<pre><code>## Too few points to calculate an ellipse</code></pre>
<pre><code>## Warning: Removed 1 rows containing missing values (geom_path).</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<pre class="r"><code># show the NMDS coordination num
head(data.sm)</code></pre>
<pre><code>##                      NMDS1       NMDS2                DT
## AH105OBT1S2R3 -0.041086594 -0.00177474    11/8/2016 Soil
## AH106OBT1S2R1  0.002871823 -0.03935155    11/8/2016 Soil
## AH109MT2      -1.558714093  0.01655423 11/12/2016 Manure
## AH110MT2      -1.569347582 -0.00676236 11/12/2016 Manure
## AH111MT2      -1.561752738 -0.02277035 11/12/2016 Manure
## AH112MT2      -1.550515008  0.03666997 11/12/2016 Manure</code></pre>
</div>
<div id="diversity" class="section level2">
<h2>Diversity</h2>
<p>You may want to compare diversity between treatment types. Pick your favorite diversity measure(s) and use the following code to generate boxplots. I also use ANOVA to test for differences in diversity between samples.</p>
<pre class="r"><code>#Diversity Boxplots
p &lt;- plot_richness(pruned, x=&quot;DT&quot;, measures=c(&quot;Shannon&quot;))
ggplot(p$data, aes(x = DT, y = value, color = NULL))+geom_boxplot(alpha=0)+geom_jitter(width=0.2)</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<pre class="r"><code>#statistical test
divtest &lt;- aov(value ~ DT,p$data)
summary(divtest)</code></pre>
<pre><code>##             Df Sum Sq Mean Sq F value Pr(&gt;F)    
## DT           3 27.192   9.064   187.3 &lt;2e-16 ***
## Residuals   66  3.195   0.048                   
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<pre class="r"><code>TukeyHSD(divtest)</code></pre>
<pre><code>##   Tukey multiple comparisons of means
##     95% family-wise confidence level
## 
## Fit: aov(formula = value ~ DT, data = p$data)
## 
## $DT
##                                                diff        lwr         upr
## 11/15/2016 Manure Line-11/12/2016 Manure  2.0125619  1.4043879  2.62073598
## 11/15/2016 Soil-11/12/2016 Manure         1.8459435  1.6392724  2.05261450
## 11/8/2016 Soil-11/12/2016 Manure          1.4015928  1.1804386  1.62274698
## 11/15/2016 Soil-11/15/2016 Manure Line   -0.1666185 -0.7542735  0.42103662
## 11/8/2016 Soil-11/15/2016 Manure Line    -0.6109691 -1.2038728 -0.01806549
## 11/8/2016 Soil-11/15/2016 Soil           -0.4443507 -0.6004659 -0.28823543
##                                              p adj
## 11/15/2016 Manure Line-11/12/2016 Manure 0.0000000
## 11/15/2016 Soil-11/12/2016 Manure        0.0000000
## 11/8/2016 Soil-11/12/2016 Manure         0.0000000
## 11/15/2016 Soil-11/15/2016 Manure Line   0.8774891
## 11/8/2016 Soil-11/15/2016 Manure Line    0.0409636
## 11/8/2016 Soil-11/15/2016 Soil           0.0000000</code></pre>
</div>
<div id="networks" class="section level2">
<h2>Networks</h2>
<p>Networks display associations of occurrence between the samples. You can also throw in variables, like temperature, into the network. They often display information that other graphs display better, but they can be used as a hypothesis generator (THEY CAN NOT TEST HYPOTHESIS). No one really has found a good way to utilize networks, but who knows, maybe in a few years, they will be very useful.</p>
<pre class="r"><code>#network
ig = make_network(pruned, type = &quot;samples&quot;, distance = &quot;bray&quot;, max.dist = 0.85)
plot_network(ig, pruned, color = &quot;DT&quot;, line_weight = 0.4, label = NULL)</code></pre>
<pre><code>## Warning: attributes are not identical across measure variables; they will
## be dropped</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>That was messy and does not tell us any taxonomic information. Lets only plot the top 100 abundant OTUs and plot by phyla.</p>
<pre class="r"><code># prune to just the top 100 most abundant OTUs across all samples (crude).
GP100 = prune_taxa(names(sort(taxa_sums(pruned), TRUE))[1:100], pruned)
jg = make_network(GP100, &quot;taxa&quot;, &quot;jaccard&quot;, 0.3)
plot_network(jg, pruned, &quot;taxa&quot;,color=&quot;Rank2&quot;, line_weight = 0.4, label = NULL)</code></pre>
<pre><code>## Warning: attributes are not identical across measure variables; they will
## be dropped</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<div id="heatmaps" class="section level2">
<h2>Heatmaps</h2>
<p>Heat maps are awful displays of data, but here you go.</p>
<pre class="r"><code>gpt &lt;- subset_taxa(pruned, Rank1==&quot;Bacteria&quot;)
gpt &lt;- prune_taxa(names(sort(taxa_sums(pruned),TRUE)[1:300]), gpt)
plot_heatmap(gpt, &quot;NMDS&quot;, &quot;bray&quot;, &quot;Type&quot;, &quot;Rank3&quot;, low=&quot;#66CCFF&quot;, high=&quot;#000033&quot;)</code></pre>
<pre><code>## Warning in metaMDS(veganifyOTU(physeq), distance, ...): stress is (nearly)
## zero: you may have insufficient data</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>Alternatively, you can do:</p>
<pre class="r"><code>#Another heatmap way
heatmap(otu_table(gpt))</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
</div>
<div id="calculating-abundance-differences" class="section level2">
<h2>Calculating Abundance differences</h2>
<p>So you your treatment types are different in phyla composition, but you want to know exactly what is the abundance difference of each phyla type. I use DEseq2 to calculate this. NOTE: This code can take awhile to run. Sit back and relax (and hope it works the first time).</p>
<pre class="r"><code>#This function requires the data in a different format which is why we are re-filtering the data
# filter out OTUs less than 10
filtered &lt;- filter_taxa(darte_ed_16s, function(x) sum(x) &gt; 10, TRUE)
#Get rid of small taxa and NTC
NoSmallTax &lt;- filter_taxa(filtered, function(x) sum(x) &gt; .001, TRUE)
NoNTC &lt;- prune_samples(!(rownames(sample_data(NoSmallTax)) %in% to_remove), NoSmallTax)

#Format change
diagdds = phyloseq_to_deseq2(NoNTC, ~ DT)</code></pre>
<pre><code>## converting counts to integer mode</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre class="r"><code>#manually calculate gm mean
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x &gt; 0]), na.rm=na.rm) / length(x))
}
geoMeans = apply(counts(diagdds), 1, gm_mean)
diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre class="r"><code>diagdds = DESeq(diagdds, test=&quot;Wald&quot;, fitType=&quot;parametric&quot;)</code></pre>
<pre><code>## using pre-existing size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre><code>## -- replacing outliers and refitting for 4579 genes
## -- DESeq argument &#39;minReplicatesForReplace&#39; = 7 
## -- original counts are preserved in counts(dds)</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre><code>##   Note: levels of factors in the design contain characters other than
##   letters, numbers, &#39;_&#39; and &#39;.&#39;. It is recommended (but not required) to use
##   only letters, numbers, and delimiters &#39;_&#39; or &#39;.&#39;, as these are safe characters
##   for column names in R. [This is a message, not an warning or error]</code></pre>
<pre class="r"><code>#results table for two dates
res = results(diagdds, contrast=c(&quot;DT&quot;, &quot;11/8/2016 Soil&quot;, &quot;11/15/2016 Soil&quot;))
alpha = 0.01
sigtab = res[which(res$padj &lt; alpha), ]
sigtab = cbind(as(sigtab, &quot;data.frame&quot;), as(tax_table(darte_ed_16s)[rownames(sigtab), ], &quot;matrix&quot;))
head(sigtab)</code></pre>
<pre><code>##           baseMean log2FoldChange     lfcSE      stat       pvalue
## Otu00001 114.52556       1.235233 0.4249514  2.906763 3.651892e-03
## Otu00002  94.23483      -2.856171 0.4205021 -6.792288 1.103688e-11
## Otu00003  52.98256      -3.298384 0.4071907 -8.100343 5.480442e-16
## Otu00005  43.87833      -3.836694 0.4524096 -8.480576 2.240818e-17
## Otu00006  40.80481      -3.345396 0.3999746 -8.364020 6.061672e-17
## Otu00012  29.58220      -3.927006 0.4465412 -8.794274 1.439756e-18
##                  padj    Rank1                 Rank2                 Rank3
## Otu00001 7.916122e-03 Bacteria            Firmicutes            Clostridia
## Otu00002 3.483109e-10 Bacteria       Verrucomicrobia        Spartobacteria
## Otu00003 9.800857e-14 Bacteria         Acidobacteria     Acidobacteria_Gp6
## Otu00005 8.014658e-15 Bacteria         Acidobacteria     Acidobacteria_Gp6
## Otu00006 1.626044e-14 Bacteria         Acidobacteria    Acidobacteria_Gp16
## Otu00012 1.544858e-15 Bacteria Bacteria_unclassified Bacteria_unclassified
##                                Rank4                       Rank5
## Otu00001               Clostridiales            Clostridiaceae_1
## Otu00002 Spartobacteria_unclassified Spartobacteria_unclassified
## Otu00003                         Gp6            Gp6_unclassified
## Otu00005                         Gp6            Gp6_unclassified
## Otu00006                        Gp16           Gp16_unclassified
## Otu00012       Bacteria_unclassified       Bacteria_unclassified
##                                Rank6
## Otu00001   Clostridium_sensu_stricto
## Otu00002 Spartobacteria_unclassified
## Otu00003            Gp6_unclassified
## Otu00005            Gp6_unclassified
## Otu00006           Gp16_unclassified
## Otu00012       Bacteria_unclassified</code></pre>
<pre class="r"><code>#summary of sig changes
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Rank2, function(x) max(x))
x = sort(x, TRUE)
sigtab$Rank2 = factor(as.character(sigtab$Rank2), levels=names(x))

# Genus order
#x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
#x = sort(x, TRUE)

# Only want phyla changes
y &lt;- sigtab[c(&quot;log2FoldChange&quot;,&quot;Rank2&quot;)]

#Plot Phylum diff
ggplot(y, aes(x=Rank2, y=log2FoldChange))+geom_boxplot()</code></pre>
<p><img src="community_tutorial_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
</div>
<div id="rda" class="section level2">
<h2>RDA</h2>
<p>This tests for explanatory power of environmental variables. I don’t trust this that much, especially for my use - antibiotic resistant genes. However, it can generate hypothesizes to test in other experiments.</p>
<pre class="r"><code>#RDA (this is currently broken, and will be updated when Ed gets the new data)

#import gene abundance data
RDAdata.gene &lt;- read.csv(&quot;~/Desktop/RDA r .csv&quot;, row.names=1)
env &lt;- RDAdata.gene[, c(&#39;ermb&#39;,&#39;ermc&#39;)]  # select only two explanatory variables


spe &lt;- t(otu_table(darte_ed_16s_filter_re_g))


spe.log &lt;- log1p (spe)  # species data are in percentage scale which is strongly rightskewed, better to transform them
spe.hell &lt;- decostand (spe.log, &#39;hell&#39;)  # we are planning to do tb-RDA, this is Hellinger pre-transformation
tbRDA &lt;- rda (spe ~ ermb+ermc, data = env)  # calculate tb-RDA with two explanatory variables
tbRDA


plot(tbRDA,display=c(&quot;species&quot;,&quot;bp&quot;),type=&quot;text&quot;,scaling=-1)

anova.cca(tbRDA, by=&quot;term&quot;)
anova.cca(tbRDA, by=&quot;margin&quot;)
anova.cca(tbRDA, by=&quot;axis&quot;)
permutest(tbRDA)

#RDA for indv species
otu &lt;- spe.hell[, c(&#39;Otu00001&#39;)]
spRDA &lt;- rda (otu ~ ermb + ermc, data = env)
spRDA
anova.cca(spRDA, by=&quot;term&quot;)
permutest(spRDA)</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
